diff --git a/node_modules/@ephox/bedrock-server/lib/main/ts/bedrock/compiler/Webpack.js b/node_modules/@ephox/bedrock-server/lib/main/ts/bedrock/compiler/Webpack.js
index 0a33f63..c3c6926 100644
--- a/node_modules/@ephox/bedrock-server/lib/main/ts/bedrock/compiler/Webpack.js
+++ b/node_modules/@ephox/bedrock-server/lib/main/ts/bedrock/compiler/Webpack.js
@@ -38,7 +38,7 @@ const webpackSharedRules = webpackRemap.concat([
         enforce: 'pre'
     },
     {
-        test: /\.(html|htm|css|bower|hex|rtf|xml|yml)$/,
+        test: /\.(html|htm|css|bower|hex|rtf|xml|yml|svg)$/i,
         use: ['raw-loader']
     }
 ]);
